### ISSUE TITLE
Hierarchy of Encoders Not Working

### ISSUE THREAD

**Issue by rajicon** [CyberZHG/keras-transformer#32]
I'm trying to encode multiple sentences, into one embedding, so first I use a self attention encoder on each word of a sentence for each sentence, and then I use another self attention encoder on each sentence representation to get a final representation.  However, when I do this, it doesn't seem to work well.  However, if I remove either the word level or self level self attention encoder (and replace it with just mean/sum of the embeddings, it does work.  I cannot seem to figure out where the problem is, do you have any theories of what could be going wrong?

Here is (a simplified) version of my model:

    def build_context_encoder(self):
        context = Input(shape=(self.max_num_words_per_context, ), dtype=tf.int64) #num_context x max_num_words_per_context #put back

      #set up word emb
      word_embs = np.zeros((self.vocab_size+1, self.word_emb_dim))  #MASK

      for word in self.word2id_dict:
            ind = self.word2id_dict[word]
            if ind > largest_ind:
                largest_ind = ind
            w_emb = self.word_embedding_model.wv[word]
            word_embs[ind] = w_emb
        

      emb_layer  = EmbeddingRet(
            input_dim=self.vocab_size+1, #is this correct?
            output_dim=self.word_emb_dim,
            mask_zero=True, #is this what we want?  Think about this
            weights=[word_embs],  #self.embedding_matrix,
            trainable=False,
            name='Encoder-Token-Embedding',
        )   
      emb = emb_layer(context) 
      encoder_embed = emb[0] 
      sa_context_encoder = transformer.get_encoders(self.encoder_num, encoder_embed, self.head_num, self.hidden_dim) #should output batch x num_contexts x emb_size
      context_emb = SumInternal()(sa_context_encoder)
      context_encoder = Model(inputs=[context], outputs=context_emb)
      return context_encoder

    def build_model(self): 
        contexts = Input(shape=(self.max_num_context, self.max_num_words_per_context, ), dtype=tf.int64) #num_context x num_words

    
      context_embs = TimeDistributed(self.context_encoder)(contexts)

      context_embs = ZeroVectorMasker()(context_embs)  #READD

      aggregator_encoder_out = transformer.get_encoders(self.encoder_num, context_embs, self.head_num, self.hidden_dim)

      final_estimate = MeanInternal()(aggregator_encoder_out)
      model = Model(inputs=[subwords, contexts], outputs=final_estimate)
      model.compile(loss='mean_squared_error', optimizer='adam', metrics=[])
      model.summary()
      return model
